### PR TITLE
Grt: Add -drc_report_iter_step option

### PR DESCRIPTION
**Pull request by luis201420**
Report the DRC on each iteration of overflow. By default only the final report is generated. (Discussed in #3583)

### CODE REVIEW

File: src/grt/src/GlobalRouter.tcl [hunk 1]
```diff
@@ -193,6 +193,7 @@ proc set_global_routing_random { args } {
 sta::define_cmd_args "global_route" {[-guide_file out_file] \
                                   [-congestion_iterations iterations] \
                                   [-congestion_report_file file_name] \
+                                  [-drc_report_iter_step] \
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.tcl:196]: this takes an argument eg [-drc_report_iter_step step]

File: src/grt/src/fastroute/src/utility.cpp [hunk 2]
```diff
@@ -2302,6 +2304,107 @@ int FastRouteCore::edgeShiftNew(Tree& t, int net)
...
+    final_x = route.init_x;
+  }
+
+  if (route.init_y < route.final_y) {
+    init_y = route.init_y;
+    final_y = route.final_y;
+  } else {
+    init_y = route.final_y;
+    final_y = route.init_y;
+  }
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:2326]: Simplify to:
const auto [init_x, final_x] = std::minmax(route.init_x, route.final_x);
const auto [init_y, final_y] = std::minmax(route.init_y, route.final_y);

File: src/grt/src/fastroute/src/utility.cpp [hunk 3]
```diff
@@ -2302,6 +2304,107 @@ int FastRouteCore::edgeShiftNew(Tree& t, int net)
...
+
+  int llX = init_x - (tile_size_ / 2);
+  int llY = init_y - (tile_size_ / 2);
+
+  int urX = final_x + (tile_size_ / 2);
+  int urY = final_y + (tile_size_ / 2);
+
+  if ((x_grid_max_ - urX) / tile_size_ < 1) {
+    urX = x_grid_max_;
+  }
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:2336]: Is this just
```
urX = std::min(urX, x_grid_max_);
```
or is it something more complex?  Its a bit hard to follow. 

Why is nothing similar needed with llX?

File: src/grt/src/fastroute/src/utility.cpp [hunk 4]
```diff
@@ -2302,6 +2304,107 @@ int FastRouteCore::edgeShiftNew(Tree& t, int net)
...
+  if ((y_grid_max_ - urY) / tile_size_ < 1) {
+    urY = y_grid_max_;
+  }
+
+  odb::Point lower_left = odb::Point(llX, llY);
+  odb::Point upper_right = odb::Point(urX, urY);
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:2342]: const

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 5]
```diff
@@ -1167,6 +1168,11 @@ NetRouteMap FastRouteCore::run()
     }
 
     last_total_overflow = total_overflow_;
+
+    // generate DRC report each interval
+    if (drc_report_iter_step_ && i % drc_report_iter_step_ == 0) {
+      saveCongestion();
```
Comment (maliberty) [inline src/grt/src/fastroute/src/FastRoute.cpp:1174]: Is this overwriting the same file each time?
(@oharboe is that what you want?)

File: src/grt/src/GlobalRouter.tcl (hunk 1 above)
Comment (luis201420) [reply src/grt/src/GlobalRouter.tcl]: The argument was added

File: src/grt/src/fastroute/src/utility.cpp (hunk 2 above)
Comment (luis201420) [reply src/grt/src/fastroute/src/utility.cpp]: The change was applied

File: src/grt/src/fastroute/src/utility.cpp (hunk 3 above)
Comment (luis201420) [reply src/grt/src/fastroute/src/utility.cpp:2320]: I used the same GlobalRouter::globalRoutingToBox operation, 
I think it's when the last tile is less width than the `tile_size`

File: src/grt/src/fastroute/src/utility.cpp (hunk 4 above)
Comment (luis201420) [reply src/grt/src/fastroute/src/utility.cpp]: The change was applied

File: src/grt/src/fastroute/src/FastRoute.cpp (hunk 5 above)
Comment (luis201420) [reply src/grt/src/fastroute/src/FastRoute.cpp]: I modified to generate a file for each interval (eg. `congestion-step.rpt` where `step` is the iteration number)

File: src/grt/src/fastroute/include/FastRoute.h [hunk 6]
```diff
@@ -99,7 +99,7 @@ class FastRouteCore
   ~FastRouteCore();
 
   void clear();
-  void saveCongestion();
+  void saveCongestion(const int& iter = -1);
```
Comment (maliberty) [inline src/grt/src/fastroute/include/FastRoute.h:102]: Just use int - pass by value is better for primitive types.

File: src/grt/src/fastroute/src/utility.cpp [hunk 7]
```diff
@@ -2351,57 +2335,72 @@ double FastRouteCore::dbuToMicrons(int64_t dbu)
   return (double) dbu / (block->getDbUnitsPerMicron());
 }
 
-void FastRouteCore::saveCongestion()
+void FastRouteCore::saveCongestion(const int& iter)
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:2338]: const int

**Review by maliberty — CHANGES_REQUESTED** [review]

**Comment by maliberty** [comment]
@eder-matheus any comments?

File: src/grt/src/fastroute/include/FastRoute.h (hunk 6 above)
Comment (oharboe) [reply src/grt/src/fastroute/include/FastRoute.h:102]: use std::option instead of overloading int? self documenting...

File: src/grt/src/fastroute/src/utility.cpp (hunk 3 above)
Comment (eder-matheus) [reply src/grt/src/fastroute/src/utility.cpp:2320]: Yes, this is checking if the last tile is smaller than tile_size, and creating a larger guide for the last row/column of the route grid.

**Comment by eder-matheus** [comment]
@luis201420 What is the default behavior? In the https://github.com/The-OpenROAD-Project/OpenROAD/issues/3583 issue, it is required that the reports should be generated by default. If I understand correctly, this code is not generating the reports by default.

@oharboe What should be the default interval between iterations to generate the congestion report?

**Comment by oharboe** [comment]
@eder-matheus The OpenROAD tool should not change behavior, but I have discussed with @maliberty the ORFS policy on when and how to generate interim .odb files, so I suggest you discuss with him. I am not the only user who has concerns on this and Matt is a bit more on top of other inputs.

**Comment by maliberty** [comment]
I think it  should be off by default at the OR level and we can discuss the default in ORFS

File: src/grt/src/GlobalRouter.tcl [hunk 8]
```diff
@@ -193,6 +193,7 @@ proc set_global_routing_random { args } {
 sta::define_cmd_args "global_route" {[-guide_file out_file] \
                                   [-congestion_iterations iterations] \
                                   [-congestion_report_file file_name] \
+                                  [-drc_report_iter_step steps] \
```
Comment (eder-matheus) [inline src/grt/src/GlobalRouter.tcl:196]: I'm not sure if this is the best name for this option. Maybe "report_partial_congestion iteration_interval", or something like that.

Also, please add the description of this new option to the README.md file.

**Review by eder-matheus — APPROVED** [review]
LGTM.

**Review by maliberty — APPROVED** [review]
